custom_metrics/origin-trials.js

```diff
@@ -0,0 +1,169 @@
+  const version = token[0];
+  versionElem = '' + version;
+  if (version !== 2 && version !== 3) {
+    validityElem = 'Unknown version';
+    return;
+  }
+
+  // Pull the fields out of the token.
+  if (token.length < 69) {
+    validityElem = 'Token is too short';
+    return;
+  }
+  const payloadLength = new DataView(token.buffer, 65, 4).getInt32(
+    0,
+    /*littleEndian=*/ false
+  );
+  const payload = new Uint8Array(token.buffer, 69);
+  if (payload.length !== payloadLength) {
+    validityElem =
+      'Token is ' + payload.length + ' bytes; expected ' + payloadLength;
+    return;
+  }
+
+  // The version + length + payload is signed.
+  const signedData = new Uint8Array(token.buffer.slice(64));
+  signedData[0] = token[0];
+
+  // Pull the fields out of the JSON payload.
+  let json;
+  try {
+    json = utf8Decoder.decode(payload);
+  } catch (e) {
+    console.error(e);
+    validityElem = 'Invalid UTF-8';
+    return;
+  }
+
+  let obj;
+  try {
+    obj = JSON.parse(json);
+  } catch (e) {
+    console.error(e);
+    validityElem = 'Invalid JSON';
+    return;
+  }
+
+  originElem = obj.origin;
+  subdomainElem = obj.isSubdomain ? 'Yes' : 'No';
+  thirdpartyElem = obj.isThirdParty ? 'Yes' : 'No';
+  usageElem = obj.usage;
+  featureElem = obj.feature;
+  let expiry;
+  try {
+    expiry = parseInt(obj.expiry);
+  } catch (e) {
+    console.error(e);
+    validityElem = "Expiry value wasn't an integer";
+    expiryElem = obj.expiry;
+    return;
+  }
+
+  let expiryDate = new Date(expiry * 1000);
+  expiryElem = expiryDate.toLocaleString();
+  if (expiryDate < new Date()) {
+    validityElem = 'Expired';
+    return;
+  }
+
+  validityElem = 'Valid';
+
+  let origin_trial_metadata = {
+    validityElem: validityElem,
+    versionElem: versionElem,
+    originElem: originElem,
+    subdomainElem: subdomainElem,
+    thirdpartyElem: thirdpartyElem,
+    usageElem: usageElem,
+    featureElem: featureElem,
+    expiryElem: expiryElem,
+  };
+
+  return origin_trial_metadata;
+}
+
+/**
+ * @function getParameterCaseInsensitive
+ * @param {Object} object
+ * @param {string} key
+ * @returns {any} value
+ */
+function getParameterCaseInsensitive(object, key) {
+  return object[
+    Object.keys(object).find((k) => k.toLowerCase() === key.toLowerCase())
+  ];
+}
+
+let requests = $WPT_REQUESTS;
+
+let meta_tags = document.querySelectorAll('meta[http-equiv="origin-trial"]');
+
+let tokens = [];
+meta_tags.forEach((tag) => tokens.push(tag.content));
+
+requests.forEach((request) => {
+  let header = getParameterCaseInsensitive(request.response_headers, 'origin-trial');
+  if (header) {
+    tokens = tokens.concat(
+      header.replaceAll(' ', '').split(',')
+    );
+  }
+});
+
+let unique_tokens = tokens.filter(
+  (value, index, self) => self.indexOf(value) === index
+);
+
+let origin_trials = [];
+
+unique_tokens.forEach((token) => {
+  let origin_trial = validate(token);
+  origin_trial.token = token;
```

Comment:
@VictorLeP I found it's already logged [here](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/almanac.js#L221)
Can we remove it?

Reply:
Ah, sure, no use in duplicating the data. (I'm assuming it won't be too terribly difficult to extract the relevant nodes using SQL.)

Reply:
Those metatags don't cover tokens from headers, so I'll leave this one.

P.S. I think we will have a hard time decoding token values with SQL: `TextDecoder` and `atob` API are not available in UDF. So there will be an effort to make polyfills work 😑